lib/charms/tempo_k8s/v1/charm_tracing.py

```diff
@@ -172,14 +172,62 @@ def my_tracing_endpoint(self) -> Optional[str]:
 provide an *absolute* path to the certificate file instead.
 """
 
+
+def _remove_stale_otel_sdk_packages():
+    """Hack to remove stale opentelemetry sdk packages from the charm's python venv.
+
+    See https://github.com/canonical/grafana-agent-operator/issues/146 and
+    https://bugs.launchpad.net/juju/+bug/2058335 for more context. This patch can be removed after
+    this juju issue is resolved and sufficient time has passed to expect most users of this library
+    have migrated to the patched version of juju.  When this patch is removed, un-ignore rule E402 for this file in the pyproject.toml (see setting
+    [tool.ruff.lint.per-file-ignores] in pyproject.toml).
+
+    This only has an effect if executed on an upgrade-charm event.
+    """
+    import os
+
+    if os.getenv("JUJU_DISPATCH_PATH") == "hooks/upgrade-charm":
+        # Imports done within this function so it is standalone, minimal side-effects, and easy to revert later
+        import logging
+        import shutil
+
+        from importlib_metadata import distributions
+
+        otel_logger = logging.getLogger("tracing")
+        otel_logger.debug("Executing _remove_stale_otel_sdk_packages patch on charm upgrade")
+        # Find any opentelemetry_* distributions
+        otel_distributions = {}
+        for d in distributions():
+            if d._normalized_name.startswith("opentelemetry_"):
+                otel_distributions[d._normalized_name] = otel_distributions.get(
+                    d._normalized_name, []
+                ) + [d]
+        otel_logger.debug(f"Found {len(otel_distributions)} opentelemetry distributions")
+
+        for package_name, package_distributions in otel_distributions.items():
+            # If we have multiple distributions of the same name, remove any that have 0 associated files
+            if len(package_distributions) > 1:
+                for pd in package_distributions:
+                    otel_logger.debug(
+                        f"Found package {package_name} with {len(package_distributions)} entries"
+                    )
+                    if pd.files is None or len(pd.files) == 0:
+                        this_path = pd._path  # type: ignore
```

Comment:
why "this" path?

Reply:
adopted "path" as you suggested offline.  `path` feels almost like a protected keyword to me because of `pathlib.Path`, but it really isn't so agreed it shouldn't be like that.  I adopted your offline suggestion